merkle/src/proof.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains traits, structs and utilities of merkle proof
+
+use std::marker::PhantomData;
+
+use crate::*;
+
+/// Merkle proof trait of a single leaf node, for details refer to <https://www.webopedia.com/definitions/merkle-proof/>
+pub trait MerkleProof {
+    /// Hash type
+    type Hash: PartialEq;
+
+    /// Calculates the root hash
+    fn root_hash(&self) -> Option<Self::Hash>;
```

Comment:
I'm not sure if I agree with the semantics here with the optional type. If a merkle proof doesn't have a root hash that means it isn't a merkle proof and so shouldn't be implementing this trait

Reply:
Maybe I should use result here. In current impl a proof is not validated on construction and it's possible to have nodes missing along the path and fail to calculate the root hash

Reply:
fixed